### PR TITLE
Implement Split Polygon

### PR DESCRIPTION
**Pull request by bansalra**
Split a polygon via line.
The gap between the polygons is configurable.

a

### CODE REVIEW

**Comment by supersonicclay** [comment]
Nice work. This will be a useful feature. Here are some observations about the functionality:

* Draw a line outside a polygon with a hole
  * Result: It adds a new polygon that fills the hole
  * Expected: change nothing
* Draw a line outside MultiPolygon
  * Result: splits the MultiPolygon into separate Polygons
  * Expected: change nothing

I think it would be less surprising if splitting didn't add new features, but instead just added polygons to existing features. This means that splitting a `Polygon` would upgrade it to a `MultiPolygon`. But splitting a `MultiPolygon` keeps it a `MultiPolygon`.

Can it work with multiple selected polygons? I'm ok if we do this in a subsequent PR.

File: docs/api-reference/mode-handlers/overview.md [hunk 1]
```diff
@@ -127,6 +127,24 @@ User can draw a new ellipse shape `Polygon` feature by clicking two corners of b
 
 User can draw a new ellipse shape `Polygon` feature by clicking center and two corners of the ellipse.
 
+## [SplitPolygonHandler](https://github.com/uber/nebula.gl/blob/master/modules/core/src/lib/mode-handlers/split-polygon-handler.js)
```
Comment (supersonicclay) [inline docs/api-reference/mode-handlers/overview.md]: also document the new edit type

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    const modeConfig = this.getModeConfig();
+    // Default gap in between the polygon
+    const { gap = 0.001, units = 'miles' } = modeConfig;
+
+    const buffer = turfBuffer(tentativeFeature, gap, { units });
+    const updatedGeometry = turfDifference(selectedGeometry, buffer);
+    this._setTentativeFeature(null);
+    if (!updatedGeometry) {
+      // eslint-disable-next-line no-console,no-undef
+      console.warn('Canceling edit. Split Polygon erased');
```
Comment (romaxa) [inline modules/core/src/lib/mode-handlers/split-polygon-handler.js]: is it expected outcome on standard operation? do we really need console.warn here?

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    return result;
+  }
+
+  splitPolygon() {
+    const selectedGeometry = this.getSelectedGeometry();
+    const tentativeFeature = this.getTentativeFeature();
+    const featureIndex = this.getSelectedFeatureIndexes()[0];
+    const modeConfig = this.getModeConfig();
+    // Default gap in between the polygon
+    const { gap = 0.001, units = 'miles' } = modeConfig;
```
Comment (romaxa) [inline modules/core/src/lib/mode-handlers/split-polygon-handler.js]: is it possible to have gap = 0? why it is not default value? I would expect 2 polygons share same points if gap = 0.

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    const buffer = turfBuffer(tentativeFeature, gap, { units });
+    const updatedGeometry = turfDifference(selectedGeometry, buffer);
+    this._setTentativeFeature(null);
+    if (!updatedGeometry) {
+      // eslint-disable-next-line no-console,no-undef
+      console.warn('Canceling edit. Split Polygon erased');
+      return null;
+    }
+
+    const pieces = updatedGeometry.geometry.coordinates.map(coordinates => {
```
Comment (romaxa) [inline modules/core/src/lib/mode-handlers/split-polygon-handler.js]: will it handle source Multipolygon properly?

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js (hunk 3 above)
Comment (georgios-uber) [reply modules/core/src/lib/mode-handlers/split-polygon-handler.js]: Why default unit is not km? So 0.001 is one meter?

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js (hunk 2 above)
Comment (georgios-uber) [reply modules/core/src/lib/mode-handlers/split-polygon-handler.js]: We should have a system similar to deck.gl for logging, making a task for that.

Comment (georgios-uber) [reply modules/core/src/lib/mode-handlers/split-polygon-handler.js]: https://github.com/uber/nebula.gl/issues/136

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js (hunk 4 above)
Comment (georgios-uber) [reply modules/core/src/lib/mode-handlers/split-polygon-handler.js]: @bansalra You also need to consider polygons with holes, in which case coordinates will be an array of arrays, the first being the main polygon and the rest are the holes to punch through.

**Comment by bansalra** [comment]
a

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js (hunk 4 above)
Comment (bansalra) [reply modules/core/src/lib/mode-handlers/split-polygon-handler.js]: Updated the code.

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js (hunk 3 above)
Comment (bansalra) [reply modules/core/src/lib/mode-handlers/split-polygon-handler.js]: Default Unit is now set to 0.1 centimeter (1mm). This way the polygons can share the same point

File: modules/core/src/lib/mode-handlers/split-polygon-handler.js [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+import booleanPointInPolygon from '@turf/boolean-point-in-polygon';
+import turfDifference from '@turf/difference';
+import turfBuffer from '@turf/buffer';
+import type { Feature, Polygon } from '../../geojson-types.js';
+import type { ClickEvent, PointerMoveEvent } from '../event-types.js';
+import type { EditAction } from './mode-handler.js';
+import { ModeHandler } from './mode-handler.js';
+
+export class SplitPolygonHandler extends ModeHandler {
```
[the remark this answers is not in the document]
Comment (bansalra) [reply modules/core/src/lib/mode-handlers/split-polygon-handler.js]: Updated the test cases.

File: docs/api-reference/mode-handlers/overview.md (hunk 1 above)
Comment (bansalra) [reply docs/api-reference/mode-handlers/overview.md]: Updated

File: docs/api-reference/mode-handlers/overview.md [hunk 6]
```diff
@@ -127,6 +127,24 @@ User can draw a new ellipse shape `Polygon` feature by clicking two corners of b
...
+
+* If the first and the last click is outside the polygon, it will split the polygon
+
+* If the clicked position is inside the polygon, it will not split the polygon
+
+### ModeConfig
+
+The following options can be provided in the `modeConfig` object:
+* `gap` (optional):  `x <number>`
+  * If gap: `x` means the spacing between the polygon would be `x`.
```
[the remark this answers is not in the document]
Comment (bansalra) [reply docs/api-reference/mode-handlers/overview.md]: Updated the doc.

**Comment by georgios-uber** [comment]
Great work @bansalra

**Comment by bansalra** [comment]
Thank you @georgios-uber

**Pull request merged**